### PR TITLE
Keychain: account manager for external systems

### PR DESCRIPTION
**Pull request by hparfr**
See readme for more info

### CODE REVIEW

**Comment by lasley** [comment]
This is very dangerous IMO & I am explicitly 👎 

This is enforcing bad practice and the entire keychain can be easily compromised in the UI with code similar to the following:

```
model = env['keychain.account']

log(
    str(env['keychain.account'].search([]).read()),
)
```

Given the current security, it wouldn't even require an admin to run this exploit.

There is also no way to expire the existing key in the event of compromise, which is a 100% necessity with anything encryption.

I'm sorry that I am not providing anything helpful to this PR, but this is because [we have been toiling over this for a while](https://github.com/OCA/server-tools/issues/471) & have not been able to come up with a production viable solution at this point. 

I would not trust my data with either this module or the PoCs we have already built for this purpose, thus the 👎 . Security is no joke, and a simple disclaimer is not enough IMO.

**Comment by hparfr** [comment]
Again, the aim of this module is to store credentials like smtp server, carriers, market places, magento, ftp servers and so on and don't leak them with db dumps or search_read. 

It's credentials you need in odoo and it's credentials which are currently stored in plain text by the modules and are easily recoverable with your snippet.

With keychain module, at runtime you need to do 2 actions to decrypt a password : 
- get the record from the ORM (which is filtered with odoo's ACL)
- get the key from the config manager (which is only accessible by python)

From a security point of view : the current state with odoo out of the box is very terrible. With this keychain module, it's certainly not perfect but a little more secure.


> This is enforcing bad practice 

Which one ?

> and the entire keychain can be easily compromised in the UI with code similar to the following:

No with keychain supported modules you will get : 
- if you are not in authorized group  : access_error 
- else : an encrypted password

With other modules, you will get : 
- if you are not in authorized group  : access_error 
- else : a clear text password

>There is also no way to expire the existing key in the event of compromise, which is a 100% necessity with anything encryption.

In the event of a compromise you have to change all your passwords.
If you want key expiration or x509 certificates your PR is welcome :) I hope the current design of this module allows this kind of enhancement.

>but this is because we have been toiling over this for a while & have not been able to come up with a production viable solution at this point.

It seams your requirements (storing credit card or personal / medical data) are very different.
You need conform to norms (like PCI-DSS) and your responsibility may be engaged if you leak this kind of data : this is not the aim of this module.

**Comment by hparfr** [comment]
Responding to comments on https://github.com/OCA/server-tools/issues/622#issuecomment-266999694

>This module will be available on the App Store. I can guarantee that it will be installed on production servers by administrators that did not read the disclaimer in the ReadMe.

This module is unusable / useless if no modules consumes his services. I hope, the developer who add this module as a dependecy of his own module will read the readme. If he don't  (which is plausible) we can't do anything about it.

>I do also like the idea of a separate service. I created a Golang app recently to do some key management & other useful things, and have been thinking of integrating it into Clouder

It's a good idea to externalize the storage of sensitive data, but it may be overkill to store smtp credentials or delivery carrier accounts.

**Comment by rvalyi** [comment]
@hparfr @lasley I won't comment about the module because I'm not a security expert. That being said, usually a good way to inject secrets in an app is via environment variables.

These can be managed with tools such https://www.vaultproject.io if something more high level than editing these secrets over ssh is required.

**Review by lasley — CHANGES_REQUESTED** [review]
@rvalyi - Ohhhhh I like that. I'm going to see about making a connector for VaultProject after I play around with it a bit more.

@hparfr - I agree on the need, but this is adding a tad bit of security and also the possibility of mass data corruption if the key needs to be revoked. The tradeoff is more dangerous than it is worth IMO.

Obviously the community can override my feelings, but they are firm on this. My project was doing nearly the exact same things you are here & I abandoned it because it is flawed from the ground up.

I recommend reading into OWASP guidelines on these subjects. They explain it all much better then me, and you will notice that nearly nothing of the guidelines is covered. 

Again, I agree it isn't covered in Odoo either, but you are specifically advertising this module as a `keychain`. That name implicitly states you are managing the keys in a safe fashion.

* https://www.owasp.org/index.php/Key_Management_Cheat_Sheet
* https://www.owasp.org/index.php/Cryptographic_Storage_Cheat_Sheet
* https://www.owasp.org/index.php/Password_Storage_Cheat_Sheet

**Comment by hparfr** [comment]
Thanks @lasley for having taking time to expose your arguments.

**Comment by lasley** [comment]
@hparfr - I honestly feel real bad about blocking this & am trying to come up with a viable solution for the key management + a proposal to make this more secure. I'll update you.

**Comment by bealdav** [comment]
Hi @lasley,

Ok there are more secure ways with external and complex system, but our all customers can't pay to have an erp (aka all in one features system) and 36 connectors with external system to maintain over the time.
It depends of criticity of these keys

@hparfr proposal use a secure way, I think

>     Fernet is built on top of a number of standard cryptographic primitives. Specifically it uses:
>
>        AES in CBC mode with a 128-bit key for encryption; using PKCS7 padding.
>        HMAC using SHA256 for authentication.
>        Initialization vectors are generated using os.urandom().

It ensures all password are not clear text.

It's a v9 module. If in v10 or v11 you provide a better way to do it, then we'll migrate to the new solution, but please don't break continuous improvement of the community, it's a step in the right direction

https://en.wikipedia.org/wiki/Continual_improvement_process

:+1:

**Comment by lasley** [comment]
@bealdav - This isn't simply about security. This is also about potential data destruction with no avenue of resolve. Encryption isn't just flipping a few bits, then flipping them back with the appropriate key when needed - you also need to build a way for re-encryption of data against new keys & the revocation of old. This is standard in any application that encrypts anything, and any product without it is dangerous. Continuous improvement does not help the user of the application in the middle of a breach scenario that is faced with the decision of having to destroy all their encrypted data in order to revoke a key. Revocation is not a feature of an encryption system, it is a requirement.

**Comment by bealdav** [comment]
@lasley If security was the main point, you'd haven't choose odoo at all, don't you think ?
Here it's better secure than odoo standard practice. Nowdays any module store its password with no rule, some time in clear text.
Rome was not built in one day and here it's v9, an old version, now.
Improvement will comes after in v10
You can propose a feature to provide revocation, no problem.
@hparfr could you consider to add revocation in roadmap ?

@lasley Thanks for your contribution on security subjects. It's a real important for us to have expert like you on this matter, just let community go step by step to real good practice as you suggest.

**Comment by legalsylvain** [comment]
Hi,
I'm not expert in the security domain, but I find the design of this module very good. To steal uncrypted credentials you have to have access to the database AND the config openerp file. So I see it as something secured.
Maybe there are other possibilities, but for the time being, this is a good solution, and **better than nothing**, specially because there are actually some OCA / not OCA modules with clear credentials (not secure), or others with clear token (not very secure) so :+1: for the concept.

 @bealdav said :
> It's a v9 module. If in v10 or v11 you provide a better way to do it, then we'll migrate to the new solution, but please don't break continuous improvement of the community, it's a step in the right direction

:+1: on that point, too.

@OCA/board , @StefanRijnhart, @others, could you give your PoV ?

kind regards.

**Comment by bealdav** [comment]
ping @guewen @yvaucher

**Comment by mourad-ehm** [comment]
:+1:  because this module make one step in the right direction. It brings a solution to managing credentials for external applications which not exist now in odoo.

**Comment by sebastienbeau** [comment]
Hi, after reading twice the comment of @lasley, I also think that we will need a process for updating the key (we should be able to unencrypt the data and encrypt it again with the new key). But I do not think we should stop this PR for that. We should add this feature in the README (in the roadmap, @hparfr  can you add it?), so everyone know that this feature is missing, and we can add it later. In any case if we want to change the key an administrator can still do it by changing it manually (extracting password, change the key and setting back the password).
So :+1:

**Comment by bguillot** [comment]
I agree with @legalsylvain @bealdav @mourad-ehm 

This can be improved but we need this improvement to have something to manage credentials because this is better than what we have right now.

:+1:

**Comment by elicoidal** [comment]
I agree with @lasley and his feedback about revocation. This is a must.
Nevertheless I would not block this PR which provides an additional step toward better password management.
I wish we would have a core and complete method to encrypt password but until then :+1:

**Comment by lasley** [comment]
Everyone tossing thumbs on this - please use the Github review feature. Otherwise, it is likely your vote will not be tallied.

**Comment by bealdav** [comment]
Error on some other modules: html_image_url_extractor, server_environment, 
not keychain.

**Comment by bealdav** [comment]
ping @mourad-ehm @bguillot @sebastienbeau @elicoidal @legalsylvain @rvalyi  see lasley comment for thumbs on top of the PR

File: keychain/README.rst [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+This module allows you to store credentials of external systems.
+
+* All the crendentials are stored in one place: easier to manage and to audit.
+* Multi-account made possible without effort.
+* Store additionnal data for each account. 
+* Validation rules for additional data.
+* Have different account for different environments (prod / test / env / etc).
+
+
+By default passwords are encrypted with a key stored in odoo config.
```
Comment (elicoidal) [inline keychain/README.rst]: add a comma after default
s/odoo config/Odoo config file

File: keychain/README.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+It's far from an ideal password storage setup, but it's way better 
+than password in clear text in the database.
+It can be easily replaced with another system. See "Security" chapter below.
```
Comment (elicoidal) [inline keychain/README.rst]: s/with/by (?)

File: keychain/README.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+Accounts may be: market places (Amazon, Cdisount, ...), carriers (Laposte, UPS, ...)
```
Comment (elicoidal) [inline keychain/README.rst]: Cdiscount

File: keychain/README.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+* Have different account for different environments (prod / test / env / etc).
+
+
+By default passwords are encrypted with a key stored in odoo config.
+It's far from an ideal password storage setup, but it's way better 
+than password in clear text in the database.
+It can be easily replaced with another system. See "Security" chapter below.
+
+Accounts may be: market places (Amazon, Cdisount, ...), carriers (Laposte, UPS, ...) 
+or any third party system called from odoo.
```
Comment (elicoidal) [inline keychain/README.rst]: Odoo

File: keychain/README.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+This module is usefull for developers.
```
Comment (elicoidal) [inline keychain/README.rst]: aimed

File: keychain/README.rst [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+The logic to choose between accounts is done in dependant modules.
```
Comment (elicoidal) [inline keychain/README.rst]: "will be achieved in dependent modules"

File: keychain/README.rst [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+
+==========
+Uses cases
+==========
+
+For delivery, each warehouse send parcels with his own account to the same carrier.
```
Comment (elicoidal) [inline keychain/README.rst]: I am not sure to understand this sentence in the usage

File: keychain/README.rst [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+
+Configuration
+=============
+
+After the installation of this module, you need to add some entries in odoo's config file :
```
Comment (elicoidal) [inline keychain/README.rst]: remove column at the end of line (duplicated)

File: keychain/README.rst [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+Configuration
+=============
+
+After the installation of this module, you need to add some entries in odoo's config file : 
+(etc/openerp.cfg) :
+
+> keychain_key = fyeMIx9XVPBBky5XZeLDxVc9dFKy7Uzas3AoyMarHPA=
+
+You can generate keys with `python keychain/bin/generate_key.py`
```
Comment (elicoidal) [inline keychain/README.rst]: add a period

File: keychain/README.rst [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+* Implement validation of user entered data:
+
+.. code:: python
+
+    class LaposteAccount(models.Model):
+        # ...
+        def _roulier_laposte_validate_data(self, data):
+            return len(data.get("agencyCode") > 3)
+
+* In your code, fetch the account :
```
Comment (elicoidal) [inline keychain/README.rst]: remove space before ":"

File: keychain/README.rst [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            'password': account.get_password()
+        }
+
+
+In this example, an account is randomly picked. Usually this is set according 
+to rules specific for each client.
+
+You have to restrict user access of your methods with suspend_security().
+
+Warning: _init_data and _validate_data should be prefixed with your namespace !
```
Comment (elicoidal) [inline keychain/README.rst]: remove space before "!"

File: keychain/README.rst [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+Choose python naming function compatible name.
+
+Switching from prod to dev
+==========================
+
+You may adopt one of the following stragegies:
```
Comment (elicoidal) [inline keychain/README.rst]: "strategies"

File: keychain/README.rst [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+* store your dev accounts in production db using the dev key
+* import your dev accounts with odoo builtin methods like a data.xml (in a dedicated module).
+* import your dev accounts with your own migration/cleanup script
+* etc.
+
+Note: only the password field is unreadable without the right key, login and data fields
```
Comment (elicoidal) [inline keychain/README.rst]: s/right/proper (?)

File: keychain/README.rst [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+between accounts.
+
+Usage (for user)
+================
+
+Go to *settings / keychain*, create a record with the following 
+
+* Namespace: type of account (ie: Laposte)
+* Name : human readable label "Warehouse 1"
+* Technical Name: name used by a consumer module (like "wharehouse_1")
```
Comment (elicoidal) [inline keychain/README.rst]: "warehouse_1"

File: keychain/README.rst [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+* Login: login of the account
+* Password_clear : For entering the password in clear text (not stored unecrypted)
```
Comment (elicoidal) [inline keychain/README.rst]: "uncrypted" or "unencrypted"?

File: keychain/README.rst [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+* Password : password encrypted, unreadable without the key (in config)
+* data: a JSON string for additionnal values (additionnal config for the account, like : `{"agencyCode": "Lyon", "insuranceLevel": "R1"})`
```
Comment (elicoidal) [inline keychain/README.rst]: s/data/Data

File: keychain/README.rst [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+* Environment : usually prod or dev or blank (for all)
+
+
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}
```
Comment (elicoidal) [inline keychain/README.rst]: update the target link and remove the comments that follows

File: keychain/README.rst [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+
+.. repo_id is available in https://github.com/OCA/maintainer-tools/blob/master/tools/repos_with_ids.txt
+.. branch is "8.0" for example
+
+Known issues / Roadmap
+======================
+- Account inheritence is not supported out of the box (like define common settings for all environments)
```
Comment (elicoidal) [inline keychain/README.rst]: In general try to use lines up to 80/100 characters to improve readability (non-blocking / nice to have)

Comment (elicoidal) [reply keychain/README.rst]: out-of-the-box
s/define/defining?

File: keychain/README.rst [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+- Adapted to work with `server_environnement` modules
+- Key expiration
+- Import passwords from data.xml
+
+Security
+========
+Common sense : Odoo is not a safe place for storing sensitive data. But sometimes you don't have much other possibilities. This module is designed to store credentials of things like carrier account, smtp, api keys... but definitively not for credits cards number, medical records et alia.
```
Comment (elicoidal) [inline keychain/README.rst]: s/much/any

Comment (elicoidal) [reply keychain/README.rst]: s/things/data.
Not sure what is "alia"

File: keychain/README.rst [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+
+
+By default, passwords are stored encrypted in the db using symetric encryption [Fernet : https://cryptography.io/en/latest/fernet/]. The encryption key is stored in openerp.tools.config.
+
+Threats even with this module installed :
+
+- unauthorized odoo user want to access data: access is rejected by odoo security rules
```
Comment (elicoidal) [inline keychain/README.rst]: "Odoo"

File: keychain/README.rst [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+- authorized odoo user try to access data with rpc api : he get the passwords encrypted, he can't recover because the key and the decrypted password are not exposed through rpc
```
Comment (elicoidal) [inline keychain/README.rst]: s/get/gets

File: keychain/README.rst [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+If your dev server is compromised, hacker can't decrypt your prod passwords since you have different keys between dev and prod.
+
+If you want something more secure : don't store any sensitive data in odoo, use an external system as a proxy, you can still use this module for storing all other data related to your accounts.
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/{project_repo}/issues>`_. In case of trouble, please
```
Comment (elicoidal) [inline keychain/README.rst]: please update target link

File: keychain/README.rst [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+
+Credits
+=======
+
+Akretion
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
```
Comment (elicoidal) [inline keychain/README.rst]: if you do not use the logo, no need for this line

File: keychain/README.rst [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+------------
+
+* Raphaël Reverdy <raphael.reverdy@akretion.com>
+
+Funders
+-------
+
+The development of this module has been financially supported by:
+
+* Akretion
```
Comment (elicoidal) [inline keychain/README.rst]: you might want to add a URL

File: keychain/__openerp__.py [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+    "category": "Uncategorized",
+    "website": "https://akretion.com/",
+    "author": "Akretion, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "external_dependencies": {
+        "python": [
+            'cryptography'],
+        "bin": [],
```
Comment (elicoidal) [inline keychain/__openerp__.py]: remove empty keys

File: keychain/models/keychain.py [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+        for rec in self:
+            rec.password = rec._encode_password(
+                rec.clear_password, rec.environment)
+
+    @api.model
+    def retrieve(self, domain):
+        """Search accounts for a given domain.
+
+        Environment is added by this function.
+        Use this instead of search() to benift from environment filtering.
```
Comment (elicoidal) [inline keychain/models/keychain.py]: benefit

File: keychain/models/keychain.py [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+    @staticmethod
+    def _serialize_data(data):
+        return json.dumps(data)
+
+    @staticmethod
+    def _parse_data(data):
+        try:
+            return json.loads(data)
+        except ValueError:
+            raise ValidationError(_("Data not valid JSON"))
```
Comment (elicoidal) [inline keychain/models/keychain.py]: "Data contains invalid JSON format"

File: keychain/models/keychain.py [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+                "key. Unless you can recover the previous key, "
+                "this password is unreadable."
+            ))
+
+    @classmethod
+    def _get_cipher(cls, force_env=None):
+        """Return a cipher using the keys of environments.
+
+        force_env = name of the env key.
+        Usefull for encoding against one precise env
```
Comment (elicoidal) [inline keychain/models/keychain.py]: "useful"

File: keychain/models/keychain.py [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+                if key and len(key) > 0  # remove False values
+            ]
+
+        if force_env:
+            envs = [force_env]
+        else:
+            envs = cls._retrieve_env()  # ex: ('dev', False)
+        keys = _get_keys(envs)
+        if len(keys) == 0:
+            _logger.info('on declanche un warning')
```
Comment (elicoidal) [inline keychain/models/keychain.py]: "Triggering a warning"

File: keychain/models/keychain.py [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+            raise Warning(_(
+                "No 'keychain_key_%s' entries found in config file. "
+                "Use a key like : %s" % (envs[0], Fernet.generate_key())
```
Comment (elicoidal) [inline keychain/models/keychain.py]: s/like :/similar to:

**Review by elicoidal — CHANGES_REQUESTED** [review]
Mostly README review
No code review

**Comment by hparfr** [comment]
Thanks @elicoidal for your precise review.

**Comment by hparfr** [comment]
Requested changes taken in account.
This PR is now ready to be merged.

Thanks everyone for your comments and reviews.

**Comment by hparfr** [comment]
What may I do for this PR to be accepted ?
* fixed requested changes
* tests ok

**Comment by bealdav** [comment]
ping @elicoidal @lasley

**Comment by lasley** [comment]
Sorry but I cannot sign off with this potential for data loss. A PSC will need to dismiss my review for merge. There's definitely enough thumbs though and it sounds like @elicoidal was onboard with the risk, so you should be good here soon.

**Review by elicoidal — APPROVED** [review]

**Comment by elicoidal** [comment]
Can you check travis?
Can you add in the readme a disclaimer/warning with a link to this discussion?

**Comment by hparfr** [comment]
Travis is failing on other modules when database_cleanup is included
https://travis-ci.org/OCA/server-tools/builds/199916286

**Comment by lasley** [comment]
@hparfr - The build was fixed in #719, so you should be able to clear the issue here with a rebase on the current 9 🚀

**Pull request merged**